### ISSUE TITLE
`exceptRange` (or `exceptYodaRange`) option for `no-extra-parens`

### ISSUE THREAD

**Issue by brettz9** [eslint/eslint#5575]
**What version are you using?**

Version on demo.

**What did you do?**

I added the following code, adapted from the non-problematic range sample from `yoda`:

```js
/*eslint yoda: [2, "never", { "exceptRange": true }]*/

function isReddish(color) {
    return (color.hue < 60 || 300 < color.hue);
}
```

**What happened?**

A `no-extra-parens` error is reported.

**What did you expect to happen?**

While it is expected that it still reports a `no-extra-parens`, I'd think that `no-extra-parens` should have its own "exceptRange" option (or "exceptYodaRange"), so that someone wishing not to normally allow them could allow them in the case of ranges.

**Comment by BYK** [comment]
You don't need parens there though, why do you think we need to allow this? (instead of you putting the `// eslint-disable-line no-extra-parens` there)

**Comment by brettz9** [comment]
The ESLint `yoda` rule states:

> When configured with the `exceptRange` option, range tests are allowed when the comparison itself is wrapped directly in parentheses, such as those of an `if` or `while` condition

Users who do not normally want to allow extra parentheses (and thus don't want to add the `no-extra-parens` rule) will be forced to get a `no-extra-parens` warning even if they do want to make an exception for yoda range conditions.

**Comment by brettz9** [comment]
In other words, putting the disable line clutters the code. If you provide a specific option, it makes sense to me that you want to let users use it without it its effects being negated (in all cases) by config from some other rule.

**Comment by ilyavolodin** [comment]
Those kind of cross-rule exceptions would be almost impossible to maintain with auto-fixing. We would need to do complete rule cross-coupling in order to do autofixes for situations like that.

**Comment by platinumazure** [comment]
Why does yoda require parentheses for this exceptional case, anyway? The whole point of the range exception should be that the first half of an or conditional, using the same identifier or member expression on both sides, is allowed to be yoda-style to make the condition more readable as a range. Parentheses should not be enforced here- the fact that they are is a sign that this rule is not following the Single Responsibility Principle. I would champion an enhancement for removing that check, so that this rule has less of a chance to collide with no-extra-parens.

**Comment by brettz9** [comment]
I agree, @platinumazure , that it'd be nicer (and simpler) for it to work without that requirement.

**Comment by mysticatea** [comment]
I agree with @platinumazure .

**Comment by btmills** [comment]
Removing the parens requirement was brought up recently in #5388, though that was closed as working as designed without any further discussion.

There were two reasons for requiring the paren check:

1. Clarity. Consider, for the test "`x` must be an odd number between 0 and 10": `x % 2 !== 0 && 0 < x && x < 10` vs `x % 2 !== 0 && (0 < x && x < 10)`. Arguably, having the parens makes the intent of the range test more clear. (Note that `if (0 < x && x < 10) { }` by itself already passes - it doesn't require *extra* parens, just that parens are present.)
2. Ease of implementation. The AST for `x % 2 !== 0 && 0 < x && x < 10` is congruent to `(x % 2 !== 0 && 0 < x) && x < 10`. Implementing the exception for that case isn't as straightforward as it is for `x % 2 !== 0 && (0 < x && x < 10)`.

**Comment by platinumazure** [comment]
I guess I'm not really convinced. Couldn't we write a utility that will take an expression tree and determine if consecutive expressions have the same precedence, and if so, apply the check?

**Comment by btmills** [comment]
If (1) is deemed insufficient, (2) from a technical perspective certainly isn't insurmountable if someone wants to tackle it.

**Comment by platinumazure** [comment]
@btmills I can respect the virtue of clarity and I'm not opposed to having
a way of enforcing it. I just don't think the Yoda rule is the right place
to do it and that we should write a new rule that focuses on clarity of
logical expressions. But that's just my opinion.

**Comment by nzakas** [comment]
Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.

**Comment by LeonardPauli** [comment]
Just putting out a thought;

Is it really necessary to check parenWrapped if the range expression already is by its own? This would pass @btmills requirements:
1. Clarity: if expression already is by its own, as in an assignment, it's still just as clear and follows the no-extra-parens rule, while the parens gets required if there are more than one boolean operator...
2. Ease of implementation: ...which keeps congruent cases out of the picture.

```javascript
/* eslint yoda: [2, "never", { "exceptRange": true }] */
/* eslint no-extra-parens: 2 */
let a, x = 6

a =  3 < x && x < 8  // pass
a = (3 < x && x < 8) // no-extra-parens
a =  x < 3 || 8 < x  // pass
a = (x < 3 || 8 < x) // no-extra-parens

a = x%2==0 && (3 < x && x < 8) // pass, range expression is by its own
a = x%2==0 &&  3 < x && x < 8  // fail, not clear and congurency issues

if (5 < x && x < 10) {} // pass
```

*Notes:*
- Same precedence is read from left to right: `a || b || c` == `(a || b) || c`
- A range check is a `LogicalExpression(BinaryExpression, BinaryExpression)`
- If `LogicalExpression` found containing another `LogicalExpression`, the inner one needs to be parenWrapped to be considered as range check

*Implementation:*
- **If `LogicalExpression(BinaryExpression, BinaryExpression)` found without LogicalExpression as parent node type, the parenWrapped is unnecessary for it to be considered as range check**

*Relevant startpoints:*
- https://github.com/eslint/eslint/blob/master/lib/rules/yoda.js#L241
- https://github.com/eslint/eslint/blob/master/lib/rules/yoda.js#L291

*Options:*
- For those who don't use the no-extra-parens rule, and always want range checks to be parenWrapped, have an option to always require parens around ranges. If set to true by default, this whole change won't change behaviour of existing configs, while allowing the simultaneous use of both `no-extra-parens` and `yoda`.

*Alternatives:*
- Add a way to mark parens as necessary, similar to the `markVariableAsUsed(name)` for the `no-unused-vars` rule.

Hope this is helpful, would really like to be able to use `no-extra-parens` and `yoda` simultaneously. If you find this solution good, I might be able to try to implement it :)

**Comment by btmills** [comment]
@LeonardPauli wow, that's very thorough! Would you mind posting that with some context as a new issue and linking this one? Having new issues helps to avoid conversations on older issues getting lost deep in the stack.

I like the idea of relaxing the paren requirement in `yoda` if a range expression is by itself :+1: We'll have to discuss on the new issue, but I want to say that could be done as a semver-minor change.

There's another case to consider in the nested case when swapping the outer `&&` for `||`:

```js
a = x%2==0 || (3 < x && x < 8)
```

Right now this fails `no-extra-parens` since precedence makes this identical to

```js
a = x%2==0 || 3 < x && x < 8
```

Any ideas about what to do there? 🤔 

To see how someone else handled this, and since this is a somewhat styling question, I tried various combinations of these in Prettier. It looks like Prettier might be special-casing range expression to wrap them in parens even when not strictly necessary according to precedence?
